---
title: "Characterisation_of_Psoralen_signature"
author: "Sigurgeir Ã“lafsson"
date: "6/22/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This document describes the various analyses carried out to characterise a mutational signature we identify and attribute to psoralen exposure. 

To enable detailed characterisation of the signature, we performed whole genome sequencing of 16 microbiopsies which showed clear evidence of the signature in the whole-exome data. The terms psoralen and PUVA exposure are sometimes used interchangibly throughout this document. Originally we attributed the signature to PUVA exposure and the names of some of the directories and early files reflect that. The "PUVA-signature" and the "Psoralen-signature" are one and the same.

```{r, warning=F, message=F}
.libPaths("/lustre/scratch119/humgen/projects/psoriasis/R_packages_farm5_R4.1.0_install/")

library("TxDb.Hsapiens.UCSC.hg38.knownGene")
library(grid)
library(gridExtra)
library(MutationalPatterns)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg38"
library(ref_genome, character.only = TRUE)
library(ggplot2)
library(reshape2)
library(ggsignif)
options(stringsAsFactors = F)

## Define a few variables
genes_hg38 <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
sample_meta <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/sample_info/sample_meta_wgs.txt", h=T, stringsAsFactors = F)
binomial_dir="/lustre/scratch119/humgen/projects/psoriasis/binomial_filters/"
repl_timing_dir="/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/07_signature_extraction/replication_timing/"
puva_dir="/lustre/scratch119/humgen/projects/psoriasis/signature_extraction/puva_characterization/"

## Read in the data
sample_names <- sample_meta$sampleID
patient_list <- sample_meta$patient_ID
vcf_files <- paste(puva_dir, sample_names, ".vcf", sep="")
grl <- read_vcfs_as_granges(vcf_files, sample_names, ref_genome, predefined_dbs_mbs=T)

```

## Mutational profiles of psoralen exposed samples

The mutational spectra of skin samples are typically dominated by the C>T mutations which characterise COSMIC signatures SBS7a and SBS7b. We can see that the samples highlighted here show evidence of some UV-exposure but their mutation spectra are dominated by T>A, T>C and T>G mutations at ApT sites. 

```{r, fig.width=10, fig.height=6}
type_occurrences <- mut_type_occurrences(grl, ref_genome)
plot_spectrum(type_occurrences, CT = TRUE, 
              indv_points = TRUE, legend = T, by=patient_list)
```

```{r, fig.width=10, fig.height=10}
mut_mat <- mut_matrix(vcf_list = grl, ref_genome = ref_genome)
plot_96_profile(mut_mat[,c(1:4)])
plot_96_profile(mut_mat[,c(5:8)])
plot_96_profile(mut_mat[,c(9:16)])
```

We may be interested in knowing if the trinucleotide model, which is classically used to describe mutational signatures, is sufficient to describe the context-dependence of this mutational process. We find that it is not so. The effect of sequence context extends beyond the trinucleotide model. 

```{r, fig.width=10, fig.height=10}
mut_mat_ext_context <- mut_matrix(grl, ref_genome, extension = 2)
plot_profile_heatmap(mut_mat_ext_context, by=patient_list)
```

## Indel spectra

We can compare the mutation spectra of indels from clones showing evidence of Psoralen exposure vs not. Because no non-exposed samples were whole-genome sequenced, this comparison will be done using the WES data. 

I classify patients as psoralen-exposed if any clone from the patient has more than 100 psoralen-associated mutations. We then pool indels across patients in both types. This is because there are too few indels in individual clones (and even across individual patients) to do meaningful signature extraction. 

```{r}
patients_wes <- read.table("/lustre/scratch119/humgen/projects/psoriasis/signature_extraction/hdp/snv/patient_list.txt")
patients_wes <- patients_wes$V1[patients_wes$V1!="patient25"] ## Removed as no microbiopsies from lesional skin passed qc.

indels <- data.frame()
for(pat in patients_wes) {
  pat_indels <- read.table(paste(binomial_dir, pat, "/", pat, "_genotype_indels.txt", sep=""), h=T)
  pat_indels$patientID <- pat
  indels <- rbind(indels, data.frame(rownames(pat_indels), pat_indels$patientID))
}

colnames(indels) <- c("mutationID", "patientID")
indels$Chr <- unlist(strsplit(indels$mutationID, split=":"))[c(T,F,F,F)]
indels$Pos <- unlist(strsplit(indels$mutationID, split=":"))[c(F,T,F,F)]
indels$Ref <- unlist(strsplit(indels$mutationID, split=":"))[c(F,F,T,F)]
indels$Alt <- unlist(strsplit(indels$mutationID, split=":"))[c(F,F,F,T)]

Grange_indels <- makeGRangesListFromDataFrame(indels, split.field = "patientID", keep.extra.columns = T, ignore.strand = T, seqnames.field = "Chr", 
                                                start.field = "Pos", end.field = "Pos")
GenomeInfoDb::genome(Grange_indels) = 'hg38'
indel_grl <- get_indel_context(Grange_indels, ref_genome)
indel_counts <- count_indel_contexts(indel_grl)
indel_counts <- data.frame(indel_counts)

## Combine the count for Psoralen exposed vs non-exposed patients
clone_sigs <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/manuscript_data_and_figures/Supplementary_material/Supplementary_Table3_clone_mutationBurden.txt", h=T)
x <- clone_sigs[clone_sigs$PUVA>100,]
psoralen_exp_patients <- unique(unlist(strsplit(x$CloneID, spli="_"))[c(T,F)])

exp <- indel_counts[,colnames(indel_counts) %in% psoralen_exp_patients]
nexp <- indel_counts[,!(colnames(indel_counts) %in% psoralen_exp_patients)]

sums <- data.frame(Exposed=rowSums(exp), NonExposed=rowSums(nexp))
plot_indel_contexts(sums, condensed = TRUE)

cos_sim(sums$Exposed, sums$NonExposed)
```


## Strand bias analyses

### Transcriptional strand bias

We can start by plotting the mutation spectra of the samples as before but now with strand information. From the plot, it is obvious that there are many more T>[ACG] mutations on the untranscribed strand. This effect is much more pronounced than the known strand bias of UV-light exposure (C>T mutations below).

```{r, fig.width=10, fig.height=10, warning=F, message=F}
genes_hg38 <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
mut_mat_s <- mut_matrix_stranded(grl, ref_genome, genes_hg38)
plot_192_profile(mut_mat_s[,1:4]) + theme_classic(base_size = 14) + theme(legend.position="top", axis.text.x = element_text(angle=90, size=8))
plot_192_profile(mut_mat_s[,5:8]) + theme_classic(base_size = 14) + theme(legend.position="top", axis.text.x = element_text(angle=90, size=8))
plot_192_profile(mut_mat_s[,9:16]) + theme_classic(base_size = 14) + theme(legend.position="top", axis.text.x = element_text(angle=90, size=8))
```

More formally, the strand_bias_test() function performs a two-sided Poisson test for the ratio between the number of mutation mapped on each strand. The difference is highly significant!
```{r}
strand_counts <- strand_occurrences(mut_mat_s)
strand_bias <- strand_bias_test(strand_counts)
strand_bias

plot_strand(strand_counts, mode = "absolute")
plot_strand_bias(strand_bias, sig_type = "p")
```

### Psoralen and transcription

To assess the relationship between psoralen-related mutagenesis and gene expression, we used gene expression data from sun-exposed skin from the [GTEx project](https://gtexportal.org/home/]). The transcription start site of each gene we got from Gencode. 

```{r}
## First read in GTEx results and merge with gencode
gtex <- read.csv("/lustre/scratch119/humgen/projects/psoriasis/resources/GTEx_skin_expression.csv", h=T)
colnames(gtex) <- c("TranscriptID", "Gene", "Skin_no_sun", "Skin_sun")

tss<-read.table("/lustre/scratch119/humgen/projects/psoriasis/resources/gencode.v27.annotation_chr_pos_strand_geneid_gene_name.gtf",h=F)
colnames(tss)<-c("Chromosome_phe","START_GENE","END_GENE","STRAND","Phenotype_ID","HGNC")
tss$TSS<-ifelse(tss$STRAND=="+",tss$START_GENE-1,tss$END_GENE-1)
```

We next extracted only protein coding genes from the nuclear genome and divided them up into 10 equally sized bins in order of ascending expression. 

```{r}
comb <- merge(gtex, tss, by.x="Gene", by.y="HGNC")
comb <- comb[order(comb$Skin_sun, decreasing = T),]
comb <- comb[comb$Chromosome_phe!="chrM",]
d <- duplicated(comb$Gene)
comb <- comb[!d,]  ## Keeps the highest expressed transcript

## Create expression bins to check the mutation rate in bins of ascending
## expression
bins <- comb[,c("Chromosome_phe", "START_GENE", "END_GENE", "Gene", "Skin_sun")]

sel_cv <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/manuscript_data_and_figures/sel_cv_dNdS_results.txt", h=T)
bins <- bins[bins$Gene %in% sel_cv$gene_name,]  ## Only consider protein coding genes. 

bins$nr <- 1:nrow(bins)
bins$expr_bin <- cut(bins$nr, breaks=10)
bins$expr_bin <- factor(bins$expr_bin, labels=c("Bin10","Bin9","Bin8","Bin7","Bin6","Bin5","Bin4","Bin3","Bin2","Bin1"))
bins$nr <- NULL
#write.table(bins, file="/lustre/scratch119/humgen/projects/psoriasis/signature_extraction/puva_characterization/GTex_expression_bins.bed", sep="\t", col.names = F, row.names = F, quote = F)

```

To calculate the mutation rate in each bin, we used the script Psoralen_expression.sh, which is available in the GitHub repository which accompanies this manuscript (see the main text).
Here, I read in and plot the results. 

```{r}
expression_res <- read.table("/lustre/scratch119/humgen/projects/psoriasis/signature_extraction/puva_characterization/Mutation_rate_pr_expr_bin.txt", h=T)
expression_res$scaled <- expression_res$Mutation_Rate/expression_res$Mutation_Rate[1]

ggplot(expression_res, aes(x=Bin, y=scaled)) + geom_bar(position="dodge",stat="identity", fill="#E24E1B", colour="black") + 
  scale_y_continuous(expand=c(0,0), limits = c(0,1.2)) + theme_bw(base_size = 14) + 
  labs(y="Scaled mutation rate \n at TpA sites", x="Genes binned by increasing expression") + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + 
  geom_hline(yintercept = 1, linetype=2)
```

### Transcription coupled repair and damage

To check for evidence of transcription coupled repair and damage, we'll continue using the GTEx data above. The procedure is as follows:

i) Compile a list of genes highly expressed in the skin. Include strand information. ii) Create 1kb bins 10kb up and downstream of the TSSs. iii) Count the number of T>A, T>C and T>G mutations at TpA sites on the transcribed strand and the number of A>T, A>G and A>C mutations at ApT sites on the untranscribed strand. If the gene is on the (+) strand, the transcribed strand is the complement of the reference strand. 

```{r}
## Create expression quintiles:
expr_quintile1 <-comb[comb$Gene %in% bins$Gene[bins$expr_bin=="Bin1" | bins$expr_bin=="Bin2"],]
expr_quintile2 <-comb[comb$Gene %in% bins$Gene[bins$expr_bin=="Bin3" | bins$expr_bin=="Bin4"],]
expr_quintile3 <-comb[comb$Gene %in% bins$Gene[bins$expr_bin=="Bin5" | bins$expr_bin=="Bin6"],]
expr_quintile4 <-comb[comb$Gene %in% bins$Gene[bins$expr_bin=="Bin7" | bins$expr_bin=="Bin8"],]
expr_quintile5 <-comb[comb$Gene %in% bins$Gene[bins$expr_bin=="Bin9" | bins$expr_bin=="Bin10"],]


## Make a bed file for each quintile with all the information necessary for the analysis.
make_bed_file <- function(geneList) {
  
  toUse <- geneList
  ## For each of these genes, I want to write 20 entries to a bed file,
  ## ten 1kb segments upstream and downstream of the TSS. 
  chr_vector <- as.character()
  start <- as.numeric()
  end <- as.numeric()
  tss <- as.character()
  strand <- as.character()
  label <- as.character()
  t <- 1
  for(i in 1:nrow(toUse)) {
    
    
    if(toUse$STRAND[i]=="-") {
      # Create 1kb bins upstream of the TSS
      for(j in 1:10) {
        chr_vector[t] <- toUse$Chromosome_phe[i]
        start[t] <- toUse$END_GENE[i] + (j-1)*1000
        end[t] <- toUse$END_GENE[i] + (j)*1000
        tss[t] <- "upstream"
        strand[t] <- "-"
        label[t] <- paste("-",j,"kb", sep="")
        t <- t+1
      }
      # Create 1kb bins downstream of the TSS
      for(k in 1:10) {
        chr_vector[t] <- toUse$Chromosome_phe[i]
        start[t] <- toUse$END_GENE[i] - (k)*1000
        end[t] <- toUse$END_GENE[i] - (k-1)*1000
        tss[t] <- "downstream"
        strand[t] <- "-"
        label[t] <- paste("+",k,"kb", sep="")
        t <- t+1
      }
    } else if(toUse$STRAND[i]=="+") {
      # Create 1kb bins upstream of the TSS
      for(j in 1:10) {
        chr_vector[t] <- toUse$Chromosome_phe[i]
        start[t] <- toUse$START_GENE[i] - (j)*1000
        end[t] <- toUse$START_GENE[i] - (j-1)*1000
        tss[t] <- "upstream"
        strand[t] <- "+"
        label[t] <- paste("-",j,"kb", sep="")
        t <- t+1
      }
      # Create 1kb bins downstream of the TSS
      for(k in 1:10) {
        chr_vector[t] <- toUse$Chromosome_phe[i]
        start[t] <- toUse$START_GENE[i] + (k-1)*1000
        end[t] <- toUse$START_GENE[i] + (k)*1000
        tss[t] <- "downstream"
        strand[t] <- "+"
        label[t] <- paste("+",k,"kb", sep="")
        t <- t+1
      }
    } else {
      stop("Strand info missing")
    }
  }
  
  bed_file <- data.frame(chr_vector, start, end, tss, strand, label)
  return(bed_file)
}

bed_q1 <- make_bed_file(expr_quintile1)
bed_q2 <- make_bed_file(expr_quintile2)
bed_q3 <- make_bed_file(expr_quintile3)
bed_q4 <- make_bed_file(expr_quintile4)
bed_q5 <- make_bed_file(expr_quintile5)

#write.table(bed_q5,file = "/lustre/scratch119/humgen/projects/psoriasis/signature_extraction/puva_characterization/quintile5_expr_genes_skin.bed",  sep="\t", quote = F, row.names = F, col.names = F)
```

I next use the psoralen_tcd.sh script, available on the Github page accompanying the manuscript (see main text), to count the number of mutations in each bin.

We can then read in and plot the results:

```{r, fig.height=8, fig.width=9}
for(q in c(1:5)) {
  assign(paste("quintile", q, "_Tmut_transcribed", sep=""), 
         read.table(paste("/lustre/scratch119/humgen/projects/psoriasis/signature_extraction/puva_characterization/quintile", q, "_Tmut_transc.bed", sep="")))

  assign(paste("quintile", q, "_Amut_transcribed", sep=""), 
         read.table(paste("/lustre/scratch119/humgen/projects/psoriasis/signature_extraction/puva_characterization/quintile", q, "_Amut_transc.bed", sep="")))

}

quintile1_Amut_transcribed$quintile=quintile1_Tmut_transcribed$quintile <- "Expression quintile1"
quintile2_Amut_transcribed$quintile=quintile2_Tmut_transcribed$quintile <- "Expression quintile2"
quintile3_Amut_transcribed$quintile=quintile3_Tmut_transcribed$quintile <- "Expression quintile3"
quintile4_Amut_transcribed$quintile=quintile4_Tmut_transcribed$quintile <- "Expression quintile4"
quintile5_Amut_transcribed$quintile=quintile5_Tmut_transcribed$quintile <- "Expression quintile5"

quintile1_Amut_transcribed$type=quintile2_Amut_transcribed$type=quintile3_Amut_transcribed$type=quintile4_Amut_transcribed$type=quintile5_Amut_transcribed$type <- "A>[CGT] transcribed,\nT>[ACG] non-transcribed" 
quintile1_Tmut_transcribed$type=quintile2_Tmut_transcribed$type=quintile3_Tmut_transcribed$type=quintile4_Tmut_transcribed$type=quintile5_Tmut_transcribed$type <- "T>[ACG] transcribed,\nA>[CGT] non-transcribed" 

comb <- rbind(quintile1_Amut_transcribed,quintile1_Tmut_transcribed,quintile2_Amut_transcribed,quintile2_Tmut_transcribed,
              quintile3_Amut_transcribed,quintile3_Tmut_transcribed,quintile4_Amut_transcribed,
              quintile4_Tmut_transcribed,quintile5_Amut_transcribed,quintile5_Tmut_transcribed)

comb$V6 <- factor(comb$V6, levels=c("-10kb", "-9kb", "-8kb", "-7kb", "-6kb", "-5kb", "-4kb", "-3kb", "-2kb", "-1kb",
                                    "+1kb", "+2kb", "+3kb", "+4kb", "+5kb", "+6kb", "+7kb", "+8kb", "+9kb", "+10kb"))

m <- melt(data.frame(table(comb$V6, comb$quintile, comb$type)))
m$variable <- NULL
colnames(m) <- c("DistanceFromTSS","expression", "MutationType", "MutationCount")

standardized <- as.numeric()
for(i in 1:nrow(m)) {
  denom <- m$MutationCount[m$DistanceFromTSS=="-10kb" & m$MutationType==m$MutationType[i] & m$expression==m$expression[i]]
  standardized[i] <- m$MutationCount[i]/denom
}
m$standardized <- standardized

m$class <- paste(m$expression, m$MutationType)

#write.table(m, "/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/manuscript_data_and_figures/puva_tcd.txt", row.names = F, quote=F, sep="\t")

ggplot(m, aes(x=DistanceFromTSS, y=standardized, colour=expression)) + geom_point() + theme_bw(base_size = 14) +
  geom_line(aes(x=DistanceFromTSS,y=standardized, color = expression, group = class, linetype=MutationType)) + 
  labs(x="Distance from the TSS", y="Mutation rate relative to intergenic regions", colour="", linetype="") + 
  scale_y_continuous(trans="log2") + scale_colour_manual(values=c("#F79256","#FBD1A2", "#7DCFB6","#00B2CA","#1D4E89"))
  
```

To test if the lines significantly depart from a straight line, we'll use a step function. We see that the effect of transcription coupled repair is highly significant from quintile 2 onwards and the effect of transcription coupled damage is highly significant from quintile 3 onwards. 

```{r}
counts <- rbind(data.frame(table(quintile1_Amut_transcribed$V6), quintile="Q1"),
                data.frame(table(quintile2_Amut_transcribed$V6), quintile="Q2"),
                data.frame(table(quintile3_Amut_transcribed$V6), quintile="Q3"),
                data.frame(table(quintile4_Amut_transcribed$V6), quintile="Q4"),
                data.frame(table(quintile5_Amut_transcribed$V6), quintile="Q5"))
  
counts$upstream_TSS <- 1
counts$upstream_TSS[grep("-", as.character(counts$Var1))] <- 0

P_vector_tcd <- as.numeric()
for(i in 1:5) {
  null_model <- lm(Freq ~ 1, data=counts[counts$quintile==paste("Q",i, sep=""),])
  alt_model <- lm(Freq ~ I(counts$upstream_TSS[counts$quintile==paste("Q",i, sep="")]), data=counts[counts$quintile==paste("Q",i, sep=""),])

  P_vector_tcd[i] <- anova(null_model, alt_model)$"Pr(>F)"[2]
}

counts <- rbind(data.frame(table(quintile1_Tmut_transcribed$V6), quintile="Q1"),
                data.frame(table(quintile2_Tmut_transcribed$V6), quintile="Q2"),
                data.frame(table(quintile3_Tmut_transcribed$V6), quintile="Q3"),
                data.frame(table(quintile4_Tmut_transcribed$V6), quintile="Q4"),
                data.frame(table(quintile5_Tmut_transcribed$V6), quintile="Q5"))
  
counts$upstream_TSS <- 1
counts$upstream_TSS[grep("-", as.character(counts$Var1))] <- 0

P_vector_tcr <- as.numeric()
for(i in 1:5) {
  null_model <- lm(Freq ~ 1, data=counts[counts$quintile==paste("Q",i, sep=""),])
  alt_model <- lm(Freq ~ I(counts$upstream_TSS[counts$quintile==paste("Q",i, sep="")]), data=counts[counts$quintile==paste("Q",i, sep=""),])

  P_vector_tcr[i] <- anova(null_model, alt_model)$"Pr(>F)"[2]
}
P_vector_tcr

data.frame(quintile=paste("Q", c(1:5), sep=""), P_tcd=P_vector_tcd, P_tcr=P_vector_tcr)

```

### Replication strand bias

Next I want to see if there is evidence of replication strand bias. This is a bit trickier than the transcription strand bias, as the replication origins of the human genome are incompletely known. Nevertheless, there are some regions which are known to be conserved (see Methods). 

Here I use a file containing replication time data from http://www.cell.com/abstract/S0092-8674(15)01714-6. In the file, left is synonymous to the leading strand because we are always using the reference as a point of reference.
As shown below, there is an enrichment of Psoralen related mutations on the leading (left) strand. 

```{r}
repli_timing <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/07_signature_extraction/strand_asymmetries/replication_timing.repdirhg38.txt")
repli_timing$Class <- NA
repli_timing$Class[repli_timing$V5==1] <- "left"
repli_timing$Class[repli_timing$V6==1] <- "right"

repli_strand <- repli_timing[!is.na(repli_timing$Class), c("V1", "V2", "V3", "Class")]
repli_strand$Ratio <- 0
colnames(repli_strand)=c("Chr", "Start", "Stop", "Class", "Ratio")

# Store in GRanges object
repli_strand_granges <- GRanges(
  seqnames = repli_strand$Chr,
  ranges = IRanges(
    start = repli_strand$Start + 1,
    end = repli_strand$Stop
  ),
  strand_info = repli_strand$Class
)
# UCSC seqlevelsstyle
seqlevelsStyle(repli_strand_granges) <- "UCSC"
repli_strand_granges

repli_strand_granges$strand_info <- factor(repli_strand_granges$strand_info,
                                           levels = c("right", "left")
)


mut_mat_s_rep <- mut_matrix_stranded(grl, ref_genome, repli_strand_granges,
                                     mode = "replication"
)
strand_counts_rep <- strand_occurrences(mut_mat_s_rep)
strand_bias_rep <- strand_bias_test(strand_counts_rep)
strand_bias_rep

ps1 <- plot_strand(strand_counts_rep, mode = "relative")
ps2 <- plot_strand_bias(strand_bias_rep)
grid.arrange(ps1, ps2)
```

As an additional sanity check, we can use a different replication origin file, the one that comes with the MutationalPatterns package. This file is in hg19 so I first did a liftOver to convert the coordinates to hg38. We see broadly consistent results using this second source of replication origin information. 

```{r}
repli_strand <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/07_signature_extraction/strand_asymmetries/MutationalPatterns_repli_strandhg38.txt", h=T)


# Store in GRanges object
repli_strand_granges <- GRanges(
  seqnames = repli_strand$Chr,
  ranges = IRanges(
    start = repli_strand$Start + 1,
    end = repli_strand$Stop
  ),
  strand_info = repli_strand$Class
)
# UCSC seqlevelsstyle
seqlevelsStyle(repli_strand_granges) <- "UCSC"
repli_strand_granges

repli_strand_granges$strand_info <- factor(repli_strand_granges$strand_info,
                                           levels = c("right", "left")
)


mut_mat_s_rep <- mut_matrix_stranded(grl, ref_genome, repli_strand_granges,
                                     mode = "replication"
)
strand_counts_rep <- strand_occurrences(mut_mat_s_rep)
strand_bias_rep <- strand_bias_test(strand_counts_rep)
strand_bias_rep

ps1 <- plot_strand(strand_counts_rep, mode = "relative")
ps2 <- plot_strand_bias(strand_bias_rep)
grid.arrange(ps1, ps2)
```

## Replication timing analysis

To see if replication timing affects the mutation rate of psoralen exposure, we estimate the mutation rate after dividing the genome up into four bins of ascending replication timing. The code for doing this analysis is available in puva_replication_timing.sh in the github repository that accompanies this manuscript (see main text). Here I just read in and plot the results.

```{r}
rt_results <- data.frame()
for(sample in sample_names) {
  rt <- read.table(paste(puva_dir, sample, "_RT_results.txt", sep=""))
  rt$sample <- sample
  rt_results <- rbind(rt_results, rt)
}

ggplot(rt_results, aes(x=V1, y=log10(V2))) + geom_boxplot() +geom_point() +
  geom_line(aes(group=sample, colour=sample)) + theme_classic(base_size = 14) +
  labs(y="log10(Mutations per TA/AT site)", x="")
```

## Effect of psoralen exposure on the clonal structure of the tissue

Psoralen exposure (and PUVA treatment in particular) results in interstrand cross links that are toxic to cells. It is possible that high exposure may result in differences in the clonal composition of the tissue and possibly different selection pressures compared with unexposed skin. 

We test for a difference in clonality first on the level of individual microbiopsies. 

```{r}
puva_colours <- c("#FCBF49", "#40476D")
clone_sigs <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/manuscript_data_and_figures/Supplementary_material/Supplementary_Table3_clone_mutationBurden.txt", h=T)
microd_meta <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/manuscript_data_and_figures/Supplementary_material/Supplementary_Table2_microdissection_metadata.txt", h=T)
x <- clone_sigs[clone_sigs$PUVA>100,]
puva_exp_patients <- unique(unlist(strsplit(x$CloneID, spli="_"))[c(T,F)])
microd_meta$PUVA_exposed <- F
microd_meta$PUVA_exposed[microd_meta$PatientID %in% puva_exp_patients] <- T

puva_vaf_test <- wilcox.test(microd_meta$MedianVAF~microd_meta$PUVA_exposed)
puva_vaf_test

ggplot(microd_meta, aes(x=PUVA_exposed, y=MedianVAF)) + geom_boxplot(aes(fill=PUVA_exposed)) +
  theme_classic(base_size = 14) + labs(x="", y="Median VAF of microbiopsy") +
  theme(legend.position = "none") + 
  scale_x_discrete(labels=c("Not exposed", "Psoralen Exposed")) + 
  scale_fill_manual(values=puva_colours) + scale_y_continuous(limits = c(0,0.6)) +
  geom_signif(annotation = paste("P = ", formatC(puva_vaf_test$p.value, digits=2), sep=""),
              y_position = c(0.55), xmin = c(1), xmax = c(2),
              tip_length = c(0.03, 0.03))
```

Next we can look at the sharing of mutations between different microbiopsies (clones that extend across multiple biopsies). We look to see if microbiopsies from Psoralen-exposed patients are enriched among the set of biopsies which are both distant (>500 micrometers apart) and share a substantial fraction (10% here) of their mutations. We find that they are and the conclusion is robust against a range of threshold values. 

```{r}
shared_by_distance <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/manuscript_data_and_figures/shared_by_distance.txt", h=T)

ggplot(shared_by_distance, aes(x=distances, y=frac_shared)) + geom_point(aes(colour=PUVA_exposed), alpha=0.75) + 
  labs(y="Fraction of shared mutations", x="Micrometers separating microbiopsies", colour="Psoralen exposed") + 
  theme_bw(base_size = 14) + theme(legend.position = "top") + 
  scale_colour_manual(values=puva_colours) +
  annotate("rect", xmin = 500, xmax = 3000, ymin = 0.1, ymax = 1,
           alpha = .2) 


shared_by_distance$inSquare <- ifelse(shared_by_distance$distances>500 & shared_by_distance$frac_shared>0.1, "Yes", "No")
chisq.test(shared_by_distance$PUVA_exposed, shared_by_distance$inSquare, correct=T)

shared_by_distance$inSquare <- ifelse(shared_by_distance$distances>250 & shared_by_distance$frac_shared>0.1, "Yes", "No")
chisq.test(shared_by_distance$PUVA_exposed, shared_by_distance$inSquare, correct=T)

shared_by_distance$inSquare <- ifelse(shared_by_distance$distances>250 & shared_by_distance$frac_shared>0.2, "Yes", "No")
chisq.test(shared_by_distance$PUVA_exposed, shared_by_distance$inSquare, correct=T)

shared_by_distance$inSquare <- ifelse(shared_by_distance$distances>750 & shared_by_distance$frac_shared>0.2, "Yes", "No")
chisq.test(shared_by_distance$PUVA_exposed, shared_by_distance$inSquare, correct=T)

```

## Signature potential damage analysis

Is the signature particularly likely to cause any class of mutations in known drivers of squamous cell carcinoma? 

```{r}
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

hdp_puva_component <- read.table("/nfs/users/nfs_s/so11/phd/psoriasis/bsub_jupyter_lab/psoriasis/manuscript_data_and_figures/Supplementary_material/Supplementary_Table5_Psoralen_signature_hdp_component.txt")
colnames(hdp_puva_component) <- "Psoralen"

## Genes mutated in SCCs and Normal skin:
## Entrez gene IDs
gene_ids <- c(196528,171023,841,1029,2195,8085,4851,4853,8493,5925,8241,7157,8626,84962,485)

signatures = get_known_signatures(genome = "GRCh38")
signatures <- as.matrix(cbind(signatures, hdp_puva_component))

contexts <- rownames(mut_mat)
context_mismatches <- context_potential_damage_analysis(contexts, txdb, ref_genome, gene_ids)
sig_damage <- signature_potential_damage_analysis(signatures[, c("SBS7b","Psoralen")], contexts, context_mismatches)
# A normalized ratio of 2 for "stop gain" mutations, means that a signature 
# is twice as likely to cause "stop gain" mutations, compared to a completely 
# random "flat" signature. 
sig_damage[1:4,c(1,4)]
```
